test/refactoring/ConvertAsync/convert_pattern.swift

```diff
@@ -5,6 +5,305 @@ enum E : Error { case e }
+  stringTupleParam { strs, err in
+    guard let (str1, str2) = strs else { return }
+    print(str1, str2)
+    if .random(), err == nil {
+      print("yay")
+    } else {
+      print("nay")
+    }
+  }
+
+  // FALLBACK:      var strs: (String, String)? = nil
+  // FALLBACK-NEXT: var err: Error? = nil
+  // FALLBACK-NEXT: do {
+  // FALLBACK-NEXT:   strs = try await stringTupleParam()
+  // FALLBACK-NEXT: } catch {
+  // FALLBACK-NEXT:   err = error
+  // FALLBACK-NEXT: }
+  // FALLBACK-EMPTY:
+  // FALLBACK-NEXT: guard let (str1, str2) = strs else { <#return#> }
+  // FALLBACK-NEXT: print(str1, str2)
+  // FALLBACK-NEXT: if .random(), err == nil {
+  // FALLBACK-NEXT:   print("yay")
+  // FALLBACK-NEXT: } else {
+  // FALLBACK-NEXT:   print("nay")
+  // FALLBACK-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=STRING-TUPLE-RESULT %s
+  stringTupleResult { res in
+    switch res {
+    case .success((let x, let y)):
+      print(x, y)
+    case .failure:
+      print("oh no")
+    }
+  }
+  // STRING-TUPLE-RESULT:      do {
+  // STRING-TUPLE-RESULT-NEXT:   let (x, y) = try await stringTupleResult()
+  // STRING-TUPLE-RESULT-NEXT:   print(x, y)
+  // STRING-TUPLE-RESULT-NEXT: } catch {
+  // STRING-TUPLE-RESULT-NEXT:   print("oh no")
+  // STRING-TUPLE-RESULT-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=STRING-TUPLE-TYPED-RESULT %s
+  stringTupleResult { res in
+    switch res {
+    case .success((let x, let y): (String, String)):
+      print(x, y)
+    case .failure:
+      print("oh no")
+    }
+  }
+  // STRING-TUPLE-TYPED-RESULT:      do {
+  // STRING-TUPLE-TYPED-RESULT-NEXT:   let (x, y) = try await stringTupleResult()
+  // STRING-TUPLE-TYPED-RESULT-NEXT:   print(x, y)
+  // STRING-TUPLE-TYPED-RESULT-NEXT: } catch {
+  // STRING-TUPLE-TYPED-RESULT-NEXT:   print("oh no")
+  // STRING-TUPLE-TYPED-RESULT-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=MIXED-TUPLE-RESULT %s
+  mixedTupleResult { res in
+    if case .failure(let err) = res {
+      print("oh no")
+    }
+    if case .success(((let x, let y), let z)) = res {
+      print("a", x, y, z)
+    }
+    switch res {
+    case .success(let ((x, _), z)):
+      print("b", x, z)
+    case .failure:
+      print("oh no again")
+    }
+  }
+
+  // MIXED-TUPLE-RESULT:      do {
+  // MIXED-TUPLE-RESULT-NEXT:   let res = try await mixedTupleResult()
+  // MIXED-TUPLE-RESULT-NEXT:   let ((x, y), z) = res
+  // MIXED-TUPLE-RESULT-NEXT:   let ((x1, _), z1) = res
+  // MIXED-TUPLE-RESULT-NEXT:   print("a", x, y, z)
+  // MIXED-TUPLE-RESULT-NEXT:   print("b", x, z)
```

Comment:
Shouldn't this be `x1, z1`?

Could we just avoid the second pattern, similar to how for single params we just keep the first naming?

Reply:
Ah, I assume the answer is "sort of" here. Ie. it would get pretty complicated and probably not worth it - the first pattern could have _ while the second binds all, etc, etc.

Reply:
> Shouldn't this be `x1, z1`?

Good catch! TIL apparently we introduce implicit vars into case stmt bodies, so we need to walk back up to the "parent" var in the pattern to handle this.

Reply:
> Ah, I assume the answer is "sort of" here. Ie. it would get pretty complicated and probably not worth it - the first pattern could have _ while the second binds all, etc, etc.

Yeah, I'm not really convinced the extra complexity will be worth it

Reply:
Heh, and now TIL! And yeah, I agree re. the complexity.